### PR TITLE
(PUP-10150) Interleave Ruby & Puppet stacks in exception reporting

### PR DESCRIPTION
**Pull request by justinstoller**
Prior to a refactor of how `PuppetStack` worked we would print the Puppet
stack frames as part of printing the Ruby stack. This was desirable in
some cases and in others it's believed that just printing the Puppet
stack would be best.

This patch re-adds the interleaving of the Puppet stack with the Ruby
stack but as an explict action during exception printing vs as a side
effect of including the Puppet stack as part of the Ruby stack.

This is achieved by adding a new accessor, `puppetstack` to our `Error`s
that deal with the reading of files. When an exception of this kind is
created it saves a reference to the current `PuppetStack`. Exceptions that
are created from these inherit the original exception's puppetstack.
`Puppet::Util::Logging.build_exception_trace` (primarily builds traces as
arrays for `ParseErrorWithIssues`),
`Puppet::Util::Logging.format_exception` (formats traces for
non-`ParseErrorWithIssues` as strings), and their underlying usage of
`Puppet::Util.pretty_backtrace` have been refactored to honor the
puppetstack if configured and available on the exception in question.

It also adds a new setting `:puppet_trace` which will only print the
Puppet stack, without any of the Ruby stack frames for those authors
working solely on Puppet code.

### CODE REVIEW

File: lib/puppet/pops/issue_reporter.rb [hunk 1]
```diff
@@ -106,11 +102,30 @@ def self.create_exception(exception_class, emit_message, formatter, diagnostic)
...
+                  nil,
+                  diagnostic.issue.issue_code,
+                  diagnostic.arguments)
+
+    if diagnostic.exception
+      exception.set_backtrace(diagnostic.exception.backtrace)
+    end
+
+    if diagnostic.exception && diagnostic.exception.respond_to?(:puppet_stacktrace)
+      exception.puppet_stacktrace = diagnostic.exception.puppet_stacktrace
```
Comment (justinstoller) [inline lib/puppet/pops/issue_reporter.rb]: just before I pushed this up I noticed this had a typo, so it must not be necessary for the case I was testing.

need to figure out a way to verify that we don't have to do shenanigans like this at various other points in our error handling code - we like to repackage exceptions.

**Comment by Magisus** [comment]
This is targeted at master, I think it should be 5.5.x?

**Comment by justinstoller** [comment]
> This is targeted at master, I think it should be 5.5.x?

Good catch, miss-targeted the PR. Updated.

**Comment by Magisus** [comment]
Fixing hooks

**Comment by justinstoller** [comment]
Rebased and updated the commit message / pr body. I think this is ready to go now.

File: lib/puppet/defaults.rb [hunk 2]
```diff
@@ -186,6 +186,11 @@ def self.valid_file_checksum_types
           Facter.trace(value) if Facter.respond_to? :trace
         end
     },
+    :puppet_trace => {
+        :default  => false,
```
Comment (Magisus) [inline lib/puppet/defaults.rb]: Does having this default to false mimic previous behavior, or maintain the performance improvement from the previous patch?

Comment (justinstoller) [reply lib/puppet/defaults.rb]: It doesn't have an effect on either, actually. Right now the PR will interleave the ruby and puppet stack traces if the `--trace` flag is passed (rationale being the expensive part is the getting of the Ruby trace, adding the puppet trace is relatively inexpensive and doesn't dilute the signal to noise ratio much if at all), if the `--trace` flag isn't passed but the `--puppet_trace` flag _is_, then we will just output the puppet stack trace, which everyone thought was the most value for the majority of users (high signal to noise ratio for actual puppet authors). ie, using the reproducer in the ticket,
with `--trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions/fail.rb:3:in block in <top (required)>'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:174:in block (2 levels) in newfunction'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler.rb:51:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:167:in block in newfunction'
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/lib/puppet/pops/puppet_stack.rb:41:in stack'
...

```





with just `--puppet_trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/broken.pp:6
/Users/justin/Backup/code/work/puppet/broken.pp:9
/Users/justin/Backup/code/work/puppet/broken.pp:0
```

Comment (Magisus) [reply lib/puppet/defaults.rb]: Ahh I get it, thanks for clearing that up.

Comment (Magisus) [reply lib/puppet/defaults.rb]: Think you could add something to the doc string for `trace` noting that it will contain both the Ruby and the Puppet frames? I think that would help disambiguate these two.

Comment (justinstoller) [reply lib/puppet/defaults.rb]: Updated, let me know if it's still lacking.

Comment (mwaggett) [reply lib/puppet/defaults.rb]: What happens if both `--trace` and `--puppet_trace` are passed?

Comment (justinstoller) [reply lib/puppet/defaults.rb]: `--puppet_trace` is a noop then, I can add a comment about that as well!

Comment (justinstoller) [reply lib/puppet/defaults.rb]: updated

File: spec/unit/util/log_spec.rb [hunk 3]
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
...
+      Puppet::Util::Log.with_destination(destination) do
+        PuppetStackCreator.new.run(Puppet::ParseErrorWithIssue)
+      end
+      Puppet[:trace] = false
+
+      # Normal logging exepectations are still true
+      expect(logs.size).to eq(1)
+      log = logs[0]
+      expect(log.message).to_not match('/log_spec.rb')
+      expect(log.backtrace[0]).to match('/log_spec.rb')
```
Comment (mwaggett) [inline spec/unit/util/log_spec.rb]: nitpick, but above you use `.first` instead of `[0]`, if you wanna be consistent

File: spec/unit/util/log_spec.rb [hunk 4]
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
...
+
+      Puppet[:trace] = false
+      Puppet[:puppet_trace] = true
+      Puppet::Util::Log.with_destination(destination) do
+        PuppetStackCreator.new.run(Puppet::ParseError)
+      end
+      Puppet[:trace] = false
+      Puppet[:puppet_trace] = false
+
+      # Normal logging exepectations are still true
```
Comment (mwaggett) [inline spec/unit/util/log_spec.rb]:
```suggestion
      # Normal logging expectations are still true
```

Comment (mwaggett) [reply spec/unit/util/log_spec.rb]: this is in a bunch of places.. 😅

**Comment by mwaggett** [comment]
This looks sane to me, but I'm curious how you tested (besides the spec tests, which I found somewhat hard to follow in terms of what all the different test cases are actually checking).

**Comment by justinstoller** [comment]
Glen provides a reproducer manifest in the ticket. I used this with ‘bundle exec puppet apply broken.pp <trace flags>’

File: spec/unit/util/log_spec.rb [hunk 5]
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
       log = logs[0]
       expect(log.message).to_not match('/log_spec.rb')
       expect(log.backtrace).to be_a(Array)
+      expect(log.backtrace.first).to match('/log_spec.rb')
+    end
+
+    it "backtrace has interleaved PuppetStack when logging ParseErrorWithIssue" do
+      logs = []
+      destination = Puppet::Test::LogCollector.new(logs)
+      Puppet::Util::Log.newdestination(destination)
```
Comment (joshcooper) [inline spec/unit/util/log_spec.rb]: puppet's spec_helper creates a `LogCollector` and makes it available to each rspec example as `@logs` instance variable. So I think you can drop the part above, and below do `expect(@logs.size).to eq(1)`

Comment (justinstoller) [reply spec/unit/util/log_spec.rb]: Oh, nice!

**Comment by justinstoller** [comment]
Addressed Molly's spelling and consistency comments & reduced the test code using the helpers Josh pointed out. I also moved the tests from log_spec to logging_spec since they're all testing the log_exception function provided by logging. I also added some describe blocks and reworded some test titles to try and make them easier to follow. Let me know if there's more I should do!

**Comment by justinstoller** [comment]
(just pushed a fix to a typo in the first line of the commit message "how how" -> "of how")
